stubs/tensorflow/tensorflow/io/__init__.pyi

```diff
@@ -0,0 +1,103 @@
+_MemoryLevels: TypeAlias = Literal[1, 2, 3, 4, 5, 6, 7, 8, 9] | None
+
+class TFRecordOptions:
+    compression_type: _CompressionTypes | TFRecordOptions
+    flush_mode: int | None  # The exact values allowed comes from zlib
+    input_buffer_size: int | None
+    output_buffer_size: int | None
+    window_bits: int | None
+    compression_level: _CompressionLevels
+    compression_method: str | None
+    mem_level: _MemoryLevels
+    compression_strategy: int | None  # The exact values allowed comes from zlib
+
+    def __init__(
+        self,
+        compression_type: _CompressionTypes | TFRecordOptions = None,
+        flush_mode: int | None = None,
+        input_buffer_size: int | None = None,
+        output_buffer_size: int | None = None,
+        window_bits: int | None = None,
+        compression_level: _CompressionLevels = None,
+        compression_method: str | None = None,
+        mem_level: _MemoryLevels = None,
+        compression_strategy: int | None = None,
+    ) -> None: ...
+    @classmethod
+    def get_compression_type_string(cls, options: _CompressionTypes | TFRecordOptions) -> str: ...
+
+class TFRecordWriter:
+    def __init__(self, path: str, options: _CompressionTypes | TFRecordOptions | None = None) -> None: ...
+    def write(self, record: bytes) -> None: ...
+    def flush(self) -> None: ...
+    def close(self) -> None: ...
+    def __enter__(self) -> Self: ...
+    def __exit__(
+        self, exc_type: type[BaseException] | None, exc_val: BaseException | None, exc_tb: TracebackType | None
+    ) -> None: ...
+
+class FixedLenFeature(NamedTuple):
+    shape: _ShapeLike
+    dtype: _DTypeLike
+    default_value: _TensorCompatible | None = None
+
+class FixedLenSequenceFeature(NamedTuple):
+    shape: _ShapeLike
+    dtype: _DTypeLike
+    allow_missing: bool = False
+    default_value: _TensorCompatible | None = None
+
+class VarLenFeature(NamedTuple):
+    dtype: _DTypeLike
+
+class SparseFeature(NamedTuple):
+    index_key: str | list[str]
+    value_key: str
+    dtype: _DTypeLike
+    size: int | list[int]
+    already_sorted: bool = False
+
+class RaggedFeature(NamedTuple):
+    # Mypy doesn't support nested NamedTuples, but runtime they actually do use
+    # nested collections.namedtuple.
+    class RowSplits(NamedTuple):  # type: ignore[misc]
+        key: str
+
+    class RowLengths(NamedTuple):  # type: ignore[misc]
+        key: str
+
+    class RowStarts(NamedTuple):  # type: ignore[misc]
+        key: str
+
+    class RowLimits(NamedTuple):  # type: ignore[misc]
+        key: str
+
+    class ValueRowIds(NamedTuple):  # type: ignore[misc]
+        key: str
+
+    class UniformRowLength(NamedTuple):  # type: ignore[misc]
+        length: int
+    dtype: _DTypeLike
+    value_key: str | None = None
+    partitions: tuple[RowSplits | RowLengths | RowStarts | RowLimits | ValueRowIds | UniformRowLength, ...] = ()  # type: ignore[name-defined]
+    row_splits_dtype: _DTypeLike = ...
```

Comment:
This is a recent bug/regression in pyright. It passes on 1.1.295 but fails on 1.1.303. I'll try to minimize this and file a bug report to pyright.

Reply:
Hmm I can't type ignore this with current pyright config settings. We have reportUnnecessaryTypeIgnoreComment set to error in both pyrightconfig.json and pyrightconfig.stricter.json, but this only errors with pyrightconfig.stricter.json. If a line hits an error in stricter file but not looser one it becomes impossible to ignore. @Avasam

Reply:
Already fixed by pyright. I'm fine waiting til next release and updating pyright version.

Reply:
Pyright releases are pretty quick, I reckon you can probably wait for this to be fixed, and in the mean time have this PR complete while assuming it'll work.
Otherwise you could add `stubs/tensorflow/tensorflow/io/__init__.pyi` tot eh stricter configs exclusions, but I'm pretty sure the next release of Pyright will be soon enough that won't be necessary.

Reply:
I believe Eric does a new pyright release every Tuesday, so yeah, let's just wait for the new release :)